Sgx updates

kylebystrom opened the pull request:
Updates to SGX documentation and examples corresponding to the additions in Pyscf PR 945 https://github.com/pyscf/pyscf/pull/945.

kylebystrom committed:
update SGX docs

kylebystrom committed:
update SGX examples

kylebystrom committed:
update example and minor docs fix in sgx.rst

susilehtola commented on examples/sgx/00-simple_sgx.py:
```diff
@@ -14,9 +14,37 @@
+mf = scf.RHF(mol)
+mf.kernel()
+
+# Using SGX for J-matrix and K-matrix
```
```suggestion
# Using SGX for J-matrix and K-matrix, without using P-junction screening
```

susilehtola commented on examples/sgx/00-simple_sgx.py:
```diff
@@ -14,9 +14,37 @@
 
 # Using RI for Coulomb matrix while K-matrix is constructed with COS-X method
 mf.with_df.dfj = True
 mf.kernel()
```
It's a bit confusing that you say COS-X here when it's SGX in the previous example. I thought COS-X has P junctions?

kylebystrom replied:
Agreed, I fixed this.

susilehtola commented on source/user/sgx.rst:
```diff
@@ -27,27 +27,54 @@ Any :attr:`scf.hf.SCF` object :attr:`mf` can be converted to an equivalent objec
+                        set pjs = False to turn off P-junction screening.
+      warnings.warn(msg)
+    converged SCF energy = -76.0267978671524
+    converged SCF energy = -76.0267685471926
```
This looks a bit sketchy in the docs: you have 7 converged SCF energies and a five-line warning. How is one supposed to interpret this?

kylebystrom replied:
I was just showing what the output looks like for the example script, but I guess it doesn't give any useful information and adds confusion. The UserWarning is explained below anyways. Should I just not include the output?

susilehtola commented on source/user/sgx.rst:
```diff
@@ -27,27 +27,54 @@ Any :attr:`scf.hf.SCF` object :attr:`mf` can be converted to an equivalent objec
+
+:attr:`SGXHF` attribute:
+
+* :attr:`direct_scf_sgx`: Whether to use direct SCF within the SGX module, meaning that the J and K matrices are evaluated from the difference in the density matrix from the previous iteration.
```
Incremental Fock builds typically also employ resets i.e. full Fock builds every N iterations. How is N controlled in your implementation?

If you don't do full builds, you suffer from creeping in numerical error.

kylebystrom replied:
I'm not currently doing rebuilds except for when the grid level changes, and I did observe that this causes small numerical noise even for simple closed-shell organic systems (~1e-7 Eh). The direct JK evaluation in PySCF does not do rebuilds either, though I have not noticed it causing as much noise as the direct SGX version. I can add this if you want it (should be a fairly simple change), though I don't have any experience with what values of N are sensible.

susilehtola replied:
I think most programs do full rebuilds every 5 iterations or so.

kylebystrom committed:
Update examples/sgx/00-simple_sgx.py

Co-authored-by: Susi Lehtola <susi.lehtola@alumni.helsinki.fi>

kylebystrom committed:
Update 00-simple_sgx.py

kylebystrom committed:
add rebuild_nsteps attribute

kylebystrom committed:
Merge branch 'sgx_updates' of https://github.com/kylebystrom/pyscf-doc into sgx_updates

kylebystrom commented:
I added and documented a `rebuild_nsteps` attribute to control how often the JK matrix is rebuilt. I also removed the output of the example script from the SGX docs page since it didn't provide much information.
